src/x2attr.c

```diff
@@ -0,0 +1,649 @@
+IPC-D-356 files.
+
+When parsing RS274-X2, a private implementation is used to create these
+according to the standard.
+
+This is stored as a (possibly) sorted table of tuples (const char * key, const char * value).
+NOTE: sorting is by the actual
+key value, not the pointer which is essentially random.  The purpose of sorting is
+to make search more efficient if necessary, but primarily to create a consistent
+canonical attribute output order when writing RS274-X2.  This helps with testing.
+
+We could store a hash table, but that's a fairly heavyweight structure,
+so use this vector-based approach instead.  The initial link pointer
+indicates the next table to search if this one comes up empty.  That's used to
+point to the aperture attributes if this is a net attribute table.
+
+This is a variable length struct, so use g_malloc.  Most of the time, its
+size is known at creation time.  Otherwise, g_realloc is used to resize.
+We don't bother to track allocation - it's as big as it needs to be.
+
+Expect not that many different attribute keys, but a lot of objects and
+values, so a small slow struct is better than a big fast one.
+
+*/
+typedef struct x2attr_state {
+    gboolean sorted;        // TRUE when keys are sorted in canonical order.  This is
+                            // only set when _sort() called.  Reset on any modification.
+    unsigned size;          // Number of elements in following
+    struct x2attr_state * next; // Aperture attributes if this is for net (object), etc.
+    struct {
+        const char * key;   // Pointer to interned UTF-8 null terminated string.
+                            // Sort by key if 'sorted'.
+        const char * value; // Pointer to interned UTF-8 null terminated string
+    } t[1]; // 1st of 'size' elements.
+} x2attr_state_t;
+
+
+/******************
+  Local functions on xattr_state
+  
+  NOTE: const char * parameters must be interned strings!  Call with g_intern_string(foo) if in doubt.
+  
+*******************/
+
+static const char *
+_find_value(const x2attr_state_t * s, const char * key)
+{
+    unsigned i;
+    if (s)
+        for (i = 0; i < s->size; ++i)
+            if (key == s->t[i].key)
+                return s->t[i].value;
+    return NULL;
+}
+
+static int
+_find_index(const x2attr_state_t * s, const char * key)
+{
+    unsigned i;
+    if (s)
+        for (i = 0; i < s->size; ++i)
+            if (key == s->t[i].key)
+                return (int)i;
+    return -1;
+}
+
+// Like _find_value(), except will look through the linked list.
+static const char *
+_find_chained_value(const x2attr_state_t * s, const char * key)
+{
+    while (s) {
+        const char * v = _find_value(s, key);
+        if (v)
+            return v;
+        s = s->next;
+    }
+    return NULL;
+}
+
+static void
+_populate(gpointer key, gpointer value, gpointer user_data)
+{
+    x2attr_state_t * s = (x2attr_state_t *)user_data;
+    s->t[s->size].key = (const char *)key;
```

Comment:
Yes.  As it mentions in the struct definition, key and value are both interned strings, so their lifetime will persist until GLib is terminated.  Interning makes string compare very fast since just compare pointers, and I'm too lazy to use strdup all over the place.